packages/core/src/ui/InlineToolbar/index.ts

```diff
@@ -0,0 +1,165 @@
+import type { InlineToolsAdapter } from '@editorjs/dom-adapters';
+import type { InlineTool } from '../../entities/InlineTool.js';
+import { type EditorJSModel, type TextRange, Index } from '@editorjs/model';
+import { EventType } from '@editorjs/model';
+import type { Nominal } from '@editorjs/model/dist/utils/Nominal';
+import { make } from '@editorjs/dom';
+
+/**
+ * Class determines, when inline toolbar should be rendered
+ * Handles caret selection changes
+ * Forms data required in certain inline tool
+ */
+export class InlineToolbar {
+  /**
+   * Editor model instance
+   * Used for interactions with stored data
+   */
+  #model: EditorJSModel;
+
+  /**
+   * Inline tool adapter instance
+   * Used for inline tools attaching and format apply
+   */
+  #inlineToolAdapter: InlineToolsAdapter;
+
+  /**
+   * Current selection range
+   */
+  #selectionRange: TextRange | undefined = undefined;
+
+  /**
+   * Tools that would be attached to the adapter
+   */
+  #tools: InlineTool[];
+
+  /**
+   * Toolbar html element related to the editor
+   */
+  #toolbar: HTMLElement | undefined = undefined;
+
+  /**
+   * @param model - editor model instance
+   * @param inlineToolAdapter - inline tool adapter instance
+   * @param tools - tools, that should be attached to adapter
+   */
+  constructor(model: EditorJSModel, inlineToolAdapter: InlineToolsAdapter, tools: InlineTool[]) {
+    this.#model = model;
+    this.#inlineToolAdapter = inlineToolAdapter;
+    this.#tools = tools;
+
+    this.#attachTools();
+
+    this.#handleSelectionChange();
+  }
+
+  /**
+   * Handle changes of the caret selection
+   */
+  #handleSelectionChange(): void {
+    /**
+     * Listen to selection change ivents in model
+     */
+    this.#model.addEventListener(EventType.CaretManagerUpdated, (event) => {
+      const selection = window.getSelection();
+
+      /**
+       * Get current input with selection
+       */
+      if (selection) {
+        /**
+         * Do not render inline toolbar for not contenteditable elements
+         */
+        if (selection.focusNode?.nodeType !== Node.TEXT_NODE) {
```

Comment:
are you sure that this statement is really for non-contenteditable elements?

Reply:
that state checks, that selection is in `TEXT_NODE` element, it means, that only selection in text contenteditable would render toolbar (native inputs do not support inline tags, but they also would have selection, this is why we need this condition)

Reply:
put this description to the doc please